paimon-common/src/main/java/org/apache/paimon/utils/TypeUtils.java

```diff
@@ -144,11 +144,13 @@ public static Object castFromStringInternal(String s, DataType type, boolean isC
                 } else {
                     // Compatible canal-cdc
                     Float f = Float.valueOf(s);
-                    if (f.toString().length() != s.length()) {
-                        throw new NumberFormatException(
-                                s + " cannot be cast to float due to precision loss");
-                    } else {
+                    // Validate precision by comparing the rounded value
+                    if ((double) f == d || Math.abs(f - d) < Math.abs(d * 1e-6)) {
```

Comment:
why use "d * 1e-6"?
what about Double#compare?

Reply:
1. The value 1e-6 (which is 0.000001) is typically a reasonable threshold for floating-point comparisons, it's a balance between precision and practical usability.
2. This approach helps avoid rejecting values that are within the acceptable precision for FLOAT types in databases like MySQL, while still accounting for the limitations of floating-point precision.